Critical: [Phases] VaporPhase mass-basis enthalpy weights latent heat by mole fraction

## Maintainer triage (June 29, 2026)
- **Status:** Verified bug against the current `master` source in this fork.
- **Severity:** Critical.
- **Area labels:** area:phases, area:thermo.
- **Tackle batch:** P1 - critical blockers and severe unit errors.

The source review confirmed that the cited code path and failure mode are still present. The original audit details are preserved below for reproduction notes and suggested fixes.

---

## Original audit details

_Part of the model-correctness audit — see #17._

**Severity:** critical  |  **Category:** units  |  **Location:** `PharmaPy/Phases.py` lines 726-733

## What's wrong
In VaporPhase.getEnthalpy, when basis='mass' the sensible-heat term hSens is returned by the parent getEnthalpy already mass-fraction-weighted in J/kg, and getHeatVaporization(temp, basis='mass') returns the per-species latent heat deltaVap in J/kg (line 651-655). Line 730 then forms the mixture total as hVap = hSens + np.dot(deltaVap, mole_frac). For a mass-basis J/kg mixture quantity, the per-species J/kg latent heats must be weighted by MASS fractions, not mole fractions. The code uses mole_frac unconditionally regardless of the basis argument.

## Why it's incorrect
A specific (per-kg) extensive property of a mixture equals the mass-fraction-weighted sum of the per-species specific properties: h_mix[J/kg] = sum_i w_i * h_i[J/kg]. Using mole fractions to combine J/kg quantities is dimensionally/physically wrong unless all molecular weights are equal. For a water (MW 18) / heavy-component (MW 100) equimolar mixture the mole-weighted result is 1.28e6 J/kg vs the correct mass-weighted 5.99e5 J/kg, a 114% error. This wrong specific enthalpy of vapor feeds directly into energy balances (e.g. evaporator/condenser duty), corrupting the heat balance.

## Suggested regression test
Construct a VaporPhase with two species of very different MW (e.g. 18 and 100), equimolar, and call getEnthalpy(basis='mass') and getEnthalpy(basis='mole'); convert the mole-basis result to per-kg using mw_av and confirm they agree. They will currently differ. Fix: in the total_h branch use mass_frac when basis=='mass' and mole_frac when basis=='mole' (e.g. weight = mass_frac if basis=='mass' else mole_frac).

---
Verification: 2/2 adversarial reviewers confirmed (refuted: 0).

<!-- gh-triage-issue -->
## Triage: reproduced with a failing regression test

Root cause is as already diagnosed in the audit and the [independent Codex verification](https://github.com/SECQUOIA/PharmaPy/issues/18#issuecomment-4696397966): [`PharmaPy/Phases.py:730`](https://github.com/SECQUOIA/PharmaPy/blob/1c3a046/PharmaPy/Phases.py#L730) weights the per-species J/kg latent heat by `mole_frac` unconditionally, even when `basis='mass'`. Confirmed present on current base.

**Environment:** base `1c3a046` (`origin/master`), Python 3.10, numpy 1.26.4. Run: `pytest test_vaporphase_mass_latent.py`

### Failing test (self-authored, red on unchanged base)
Constructs a real `VaporPhase` from two species of very different MW (18 vs 100 g/mol), equimolar, at 350 K (below both `t_crit`), and asserts the basis-invariance `h[J/kg] == h[J/mol] * 1000 / mw_av`:

```python
def test_vaporphase_mass_basis_weights_latent_by_mass(tmp_path):
    path = tmp_path / "thermo_two_species.json"
    path.write_text(json.dumps(THERMO))  # light: mw=18, heavy: mw=100

    temp = 350.0  # K, below both critical temperatures
    phase = VaporPhase(str(path), temp=temp, moles=1.0,
                       mole_frac=[0.5, 0.5], check_input=False)

    h_mass = phase.getEnthalpy(temp=temp, basis="mass")  # J/kg
    h_mole = phase.getEnthalpy(temp=temp, basis="mole")  # J/mol
    h_mole_as_mass = h_mole / phase.mw_av * 1000.0

    assert h_mass == pytest.approx(h_mole_as_mass, rel=1e-9)
```

Red output:
```
>       assert h_mass == pytest.approx(h_mole_as_mass, rel=1e-9)
E       assert np.float64(1560462.9825554972) == 942320.581561024 ± 9.4e-04
E         Obtained: 1560462.9825554972
E         Expected: 942320.581561024 ± 9.4e-04
```

### Observed vs expected (mixture specific enthalpy at 350 K)
| Quantity | `getEnthalpy(basis='mass')` [J/kg] | correct: `getEnthalpy(basis='mole') * 1000 / mw_av` [J/kg] |
| --- | --- | --- |
| value | 1 560 463 | 942 321 |

~66% error; the mole-basis path is already correct and is used here as the reference.

### Acceptance criteria
1. **Correct value:** `getEnthalpy(basis='mass')` equals the mass-fraction-weighted mixture enthalpy, i.e. matches `getEnthalpy(basis='mole') * 1000 / mw_av` for arbitrary composition and MW disparity (the test above turns green).
2. **No regression to the correct path:** `getEnthalpy(basis='mole')` output is unchanged by the fix.
3. **Per-species path unaffected:** the `total_h=False` branch (which returns `hSens + deltaVap` without fraction weighting) is unchanged.
4. Existing test suite still passes.

Suggested next step: `gh-issue-to-pr 18`, reusing this test alongside the one-line fix (`weight = mass_frac if basis == 'mass' else mole_frac`).

--- a/PharmaPy/Phases.py
+++ b/PharmaPy/Phases.py
@@ -674,7 +674,9 @@ class VaporPhase(ThermoPhysicalManager):
         total_h : bool, optional
             If True, the total enthalpy is returned. If False, an array
             of individual enthalpy for each species is returned.
-            he default is True.
+            The default is True.
+        basis : {'mass', 'mole'}, optional
+            Basis for the returned enthalpy. The default is 'mass'.
 
         Returns
         -------
@@ -727,7 +729,8 @@ class VaporPhase(ThermoPhysicalManager):
 
         # Collect terms
         if total_h:
-            hVap = hSens + np.dot(deltaVap, mole_frac)
+            frac = mass_frac if basis == 'mass' else mole_frac
+            hVap = hSens + np.dot(deltaVap, frac)
 
         else:
             hVap = hSens + deltaVap

--- a/tests/test_phases_vapor_enthalpy.py
+++ b/tests/test_phases_vapor_enthalpy.py
@@ -1,0 +1,71 @@
+import json
+
+import numpy as np
+import pytest
+
+from PharmaPy.Phases import VaporPhase
+
+
+THERMO_TWO_SPECIES = {
+    "light": {
+        "mw": 18.0,
+        "t_crit": 650.0,
+        "rho_liq": 1000.0,
+        "cp_liq": [75.0],
+        "p_vap": [8.0, 1500.0, -40.0],
+        "delta_hvap": 40000.0,
+        "tref_hvap": 350.0,
+    },
+    "heavy": {
+        "mw": 100.0,
+        "t_crit": 700.0,
+        "rho_liq": 900.0,
+        "cp_liq": [150.0],
+        "p_vap": [8.0, 1800.0, -40.0],
+        "delta_hvap": 60000.0,
+        "tref_hvap": 350.0,
+    },
+}
+
+
+def test_vaporphase_mass_basis_weights_latent_heat_by_mass_fraction(tmp_path):
+    path = tmp_path / "thermo_two_species.json"
+    path.write_text(json.dumps(THERMO_TWO_SPECIES))
+
+    temp = 350.0
+    temp_ref = 298.15
+    mole_frac = np.array([0.5, 0.5])
+    phase = VaporPhase(
+        str(path),
+        temp=temp,
+        moles=1.0,
+        mole_frac=mole_frac,
+        check_input=False,
+    )
+
+    mw = np.array([18.0, 100.0])
+    cp_liq = np.array([75.0, 150.0])
+    delta_hvap = np.array([40000.0, 60000.0])
+    mass_frac = mole_frac * mw / np.dot(mole_frac, mw)
+
+    species_h_mole = cp_liq * (temp - temp_ref) + delta_hvap
+    expected_h_mole = np.dot(mole_frac, species_h_mole)
+
+    species_h_mass = species_h_mole * 1000.0 / mw
+    expected_h_mass = np.dot(mass_frac, species_h_mass)
+
+    h_mole = phase.getEnthalpy(temp=temp, basis="mole")
+    h_mass = phase.getEnthalpy(temp=temp, basis="mass")
+
+    assert h_mole == pytest.approx(expected_h_mole)
+    assert h_mass == pytest.approx(expected_h_mass)
+    assert h_mass == pytest.approx(h_mole * 1000.0 / phase.mw_av)
+
+    np.testing.assert_allclose(
+        phase.getEnthalpy(temp=temp, total_h=False, basis="mole"),
+        np.atleast_2d(species_h_mole),
+    )
+    np.testing.assert_allclose(
+        phase.getEnthalpy(temp=temp, total_h=False, basis="mass"),
+        np.atleast_2d(species_h_mass),
+    )

